hbase-server/src/main/java/org/apache/hadoop/hbase/master/DrainingServerTracker.java

```diff
@@ -77,36 +82,66 @@ public void serverAdded(ServerName sn) {
         }
       }
     });
+
     List<String> servers =
       ZKUtil.listChildrenAndWatchThem(watcher, watcher.getZNodePaths().drainingZNode);
-    add(servers);
+
+    if (servers != null) {
+      add(servers);
+    }
   }
 
   private void add(final List<String> servers) throws IOException {
     synchronized (this.drainingServers) {
-      this.drainingServers.clear();
+      // Clear all servers from the draining list that shouldn't stay drained. Information about
+      // whether a server should stay drained or not is added to the ZNode by the HMaster
+      this.drainingServers.removeIf(sn -> !shouldServerStayDrained(sn));
+
       for (String n : servers) {
         final ServerName sn = ServerName.valueOf(ZKUtil.getNodeName(n));
         this.drainingServers.add(sn);
         this.serverManager.addServerToDrainList(sn);
-        LOG.info("Draining RS node created, adding to list [" + sn + "]");
-
+        LOG.info("Draining RS node created, adding to list [{}]", sn);
       }
     }
   }
 
   private void remove(final ServerName sn) {
     synchronized (this.drainingServers) {
+      if (shouldServerStayDrained(sn)) {
+        LOG.info(
+          "Refusing to remove drained RS {} from the list, it's marked as permanently drained", sn);
+        return;
+      }
       this.drainingServers.remove(sn);
       this.serverManager.removeServerFromDrainList(sn);
+      LOG.info("Successfully removed drained RS {} from the list", sn);
     }
   }
 
+  private boolean shouldServerStayDrained(final ServerName sn) {
+    boolean shouldBePermanentlyDecommissioned = false;
+    String parentZnode = this.watcher.getZNodePaths().drainingZNode;
+    String node = ZNodePaths.joinZNode(parentZnode, sn.getServerName());
+
+    try {
+      byte[] rawData = ZKUtil.getData(this.watcher, node);
+      // Check if the data is present for backwards compatibility, some nodes may not have it
+      if (rawData != null && rawData.length > 0) {
+        DrainedZNodeServerData znodeData = DrainedZNodeServerData.parseFrom(rawData);
+        shouldBePermanentlyDecommissioned = znodeData.getMatchHostNameOnly();
```

Comment:
You've changed terminology between the serialized form and the in-memory form. What does this field represent? Is it "match on hostname only" or is it a claim of "permanence" ? I'm partial toward the former because there are no guarantees in this world...

More constructively though, I don't care much for the use of "permanent" in these method names. The data itself stored in ZooKeeper is considered ephemeral and not crucial for the recovery of a failed cluster. We shouldn't put anything in ZK that is so important as to be called permanent.

Try again with this method, calling it something like `honorsHostNameOnly` or similar.